src/middlewared/middlewared/plugins/ldap.py

```diff
@@ -0,0 +1,641 @@
+            ldap = self.middleware.call_sync('ldap.config')
+
+        with LDAPQuery(conf=ldap, logger=self.logger, hosts=ldap['hostname']) as LDAP:
+            ret = LDAP.get_root_DSE()
+
+        self.logger.debug(ret)
+        return ret
+
+    @private
+    async def started(self):
+        ldap = await self.config()
+        try:
+            ret = await asyncio.wait_for(self.middleware.call('ldap.get_root_DSE', ldap),
+                                         timeout=ldap['timeout'])
+        except asyncio.TimeoutError:
+            raise CallError(f'LDAP status check timed out after {ldap["timeout"]} seconds.', errno.ETIMEDOUT)
+
+        if ret:
+            await self.__set_state(DSStatus['HEALTHY'])
+        else:
+            await self.__set_state(DSStatus['FAULTED'])
+
+        return True if ret else False
+
+    @private
+    async def get_workgroup(self, ldap=None):
+        ret = None 
+        smb = await self.middleware.call('smb.config')
+        if ldap is None:
+            ldap = await self.config()
+
+        try:
+            ret = await asyncio.wait_for(self.middleware.call('ldap.get_samba_domains', ldap),
+                                         timeout=ldap['timeout'])
+        except asyncio.TimeoutError:
+            raise CallError(f'ldap.get_workgroup timed out after {ldap["timeout"]} seconds.', errno.ETIMEDOUT)
+
+        if len(ret) > 1:
+            raise CallError(f'Multiple Samba Domains detected in LDAP environment: {ret}', errno.EINVAL)
+
+        ret = ret[0] if ret else []
+
+        if ret and smb['workgroup'] != ret:
+            self.logger.debug(f'Updating SMB workgroup to match the LDAP domain name [{ret}]')
+            await self.middleware.call('datastore.update', 'services.cifs', smb['id'], {'cifs_srv_workgroup': ret[0]})
+
+        return ret
+
+    @private
+    async def __set_state(self, state):
+        await self.middleware.call('cache.put', 'LDAP_State', state.name)
+
+    @accepts()
+    async def get_state(self):
+        """
+        Check the state of the LDAP Directory Service.
+        See DSStatus for definitions of return values.
+        :DISABLED: Service is not enabled.
+        If for some reason, the cache entry indicating Directory Service state
+        does not exist, re-run a status check to generate a key, then return it.
+        """
+        ldap = await self.config()
+        if not ldap['enable']:
+            return 'DISABLED'
+        else:
+            try:
+                return (await self.middleware.call('cache.get', 'LDAP_State'))
+            except KeyError:
+                await self.started()
+                return (await self.middleware.call('cache.get', 'LDAP_State'))
+
+    @private
+    async def nslcd_cmd(self, cmd):
+        nslcd = await run(['service', 'nslcd', cmd], check=False)
+        if nslcd.returncode != 0:
+            raise CallError(f'nslcd failed to {cmd} with errror: {nscld.stderr.decode()}', errno.EFAULT)
```

Comment:
typo: `nscld`

Reply:
For the life of me I can't type those letters in the correct order twice in a row.